keps/sig-cli/2227-kubectl-default-container/README.md

```diff
@@ -0,0 +1,300 @@
+    // container. This gives users ability to preselect the most interesting container in pod.
+    if annotations := pod.Annotations; annotations != nil && len(annotations[corev1.DefaultContainerAnnotationName]) > 0 {
+      containerName := annotations[corev1.DefaultContainerAnnotationName]
+      if exists, _ := podutils.FindContainerByName(pod, containerName); exists != nil {
+        return containerName
+      } else {
+        fmt.Fprintf(p.ErrOut, "Default container name %q in annotation not found in a pod\n", containerName)
+      }
+    }
+    fmt.Fprintf(p.ErrOut, "Defaulting container name to first container %s.\n", pod.Spec.Containers[0].Name)
+
+    if p.EnableSuggestedCmdUsage {
+      fmt.Fprintf(p.ErrOut, "Use '%s describe pod/%s -n %s' to see all of the containers in this pod.\n", p.ParentCommandName, pod.Name, p.Namespace)
+    }
+  }
+
+  containerName = pod.Spec.Containers[0].Name
+  return containerName
+}
+```
+- Add test cases to make sure that the command is running with the right container. When `--container` is specified, the annotation will be ignored.
+- Validate the annotation value before using it, as container name should follow RFC 1123. If the annotation value is invalid or not found in the pod, a warning message is needed before exiting.
+- Add an Env 'KUBE_DEFAULT_CONTAINER_ANNOTATION=disabled', we will ignore the specified annotation. This Env is only for disabling this. By default, this feature should be enabled as this feature is opt-in, and it only works once user adds the specified annotation to their pods.
+- Ensure that when `kubectl.kubernetes.io/default-logs-container` is specified, we should use this annotation instead of the general one for `kubectl log` and use general annotation for other commands.
+
+
+
+**Data Command Structure:**
+
+
+**Example Command:**
+Users might specify the `kubectl.kubernetes.io/default-container` annotation in a Pod to preselect container for kubectl exec and all kubectl commands.
+
+An example Pod yaml is like below:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+ annotations:
+    kubectl.kubernetes.io/default-container: sidecar-container
+```
+
+### Test Plan
+Add a unit test for each command, testing the behavior with the annotation, without the annotation, and with the --container flag
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+As this is an opt-in feature, no gate is expected.
+- At least 2 release cycles pass to gather feedback and bug reports during
+- Documentations, add it to [well-known annotations docs](https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/)
+
+#### Beta -> GA Graduation
+
+- Gather feedback from developers and surveys
+- At least 2 release cycles pass to gather feedback and bug reports during
+- Once this feature is GA, remove env check of KUBE_DEFAULT_CONTAINER_ANNOTATION. This feature should always be enabled.
```

Comment:
Please add here that as the feature is GA, the message of deprecation of the annotation `kubectl.kubernetes.io/default-logs-container`will be removed and this annotation will stop working